File: kadanes-algorithm.py
class Solution:
    ##Complete this function
    #Function to find the sum of contiguous subarray with maximum sum.
    def maxSubArraySum(self,arr,N):
        ##Your code here
        s=0
        maxSum=arr[0]
        start=0
        poi=0
        end=0
        for i in range(N):
            s+=arr[i]
            if s>maxSum:
                start=poi
                end=i
                maxSum=s
            if s<0:
                s=0
                poi=i+1
        return maxSum
        #if find indices of sub-array
        # return [start,end]

File: DP/KadanesAlgorithm.cpp
/*
Kadane's Algorithm 
Given an array Arr[] of N integers. Find the contiguous sub-array(containing at least one number)
which has the maximum sum and return its sum.
Example 1:
Input:
N = 5
Arr[] = {1,2,3,-2,5}
Output:
9
Explanation:
Max subarray sum is 9
of elements (1, 2, 3, -2, 5) which 
is a contiguous subarray.
Example 2:
Input:
N = 4
Arr[] = {-1,-2,-3,-4}
Output:
-1
Explanation:
Max subarray sum is -1 
of element (-1)
 */
#include<iostream>
using namespace std;

int maxSum(int a[], int n){
    int ma = INT_MIN;
    int max_th = 0;
    for(int i=0; i<n; i++){
        max_th = max_th+ a[i];
        if(max_th>ma)
            ma = max_th;
        if(max_th<0)
            max_th = 0;
    }
    return ma;
}
int main(){
    int n;
    cin>>n;
    int a[n];
    for(int i=0; i<n; i++)
        cin>>a[i];
    int sum= maxSum(a, n);
    cout<<"Sum: "<<sum;
}